File: 671_easy_second-minimum-node-in-a-binary-tree/solution.js
// Definition for a binary tree node.
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Helper function to find the second minimum value
function find(node, root, ans) {
    if (node === null) {
        return;
    }

    if (node.val !== root.val) {
        if (node.val < ans.value) {
            ans.value = node.val;
        }
        return;
    } else {
        find(node.left, root, ans);
        find(node.right, root, ans);
    }
}

// Main function to find the second minimum value in the binary tree
function findSecondMinimumValue(root) {
    let ans = { value: Number.MAX_SAFE_INTEGER };

    find(root, root, ans);

    return ans.value === Number.MAX_SAFE_INTEGER ? -1 : ans.value;
}

// Example usage:
const root = new TreeNode(2, 
    new TreeNode(2), 
    new TreeNode(5, 
        new TreeNode(5), 
        new TreeNode(7)
    )
);

console.log(findSecondMinimumValue(root)); // Output should be 5

File: 671_easy_second-minimum-node-in-a-binary-tree/solution.c
#include <stdio.h>
#include <limits.h>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// Function to create a new TreeNode
struct TreeNode* createTreeNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

long ans;

// Helper function to find the second minimum value
void find(struct TreeNode* node, struct TreeNode* root) {
    if (node == NULL) {
        return;
    }

    if (node->val != root->val) {
        if (node->val < ans) {
            ans = node->val;
        }
        return;
    } else {
        find(node->left, root);
        find(node->right, root);
    }
}

// Main function to find the second minimum value in the binary tree
int findSecondMinimumValue(struct TreeNode* root) {
    ans = LONG_MAX;

    find(root, root);

    if (ans == LONG_MAX) {
        return -1;
    } else {
        return (int)ans;
    }
}

int main() {
    // Example usage:
    struct TreeNode* root = createTreeNode(2);
    root->left = createTreeNode(2);
    root->right = createTreeNode(5);
    root->right->left = createTreeNode(5);
    root->right->right = createTreeNode(7);

    int result = findSecondMinimumValue(root);
    printf("Second minimum value: %d\n", result); // Output should be 5

    return 0;
}
